dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/sms/SmsImportMapper.java

```diff
@@ -0,0 +1,346 @@
+      @Nonnull Set<String> programAttributes,
+      @Nonnull Set<String> existingAttributeValues) {
+    return Enrollment.builder()
+        .orgUnit(metadataUid(submission.getOrgUnit()))
+        .program(metadataUid(submission.getTrackerProgram()))
+        .trackedEntity(submission.getTrackedEntityInstance().getUid())
+        .enrollment(submission.getEnrollment().getUid())
+        .enrolledAt(toInstant(submission.getEnrollmentDate()))
+        .occurredAt(toInstant(submission.getIncidentDate()))
+        .status(map(submission.getEnrollmentStatus()))
+        .geometry(map(submission.getCoordinates()))
+        .attributes(
+            mapProgramAttributeValues(
+                submission.getValues(), programAttributes, existingAttributeValues))
+        .build();
+  }
+
+  /**
+   * mapProgramAttributeValues translates {@link EnrollmentSmsSubmission#getValues()} into {@link
+   * Enrollment#getAttributes()}. The tracker importer only accepts program attributes in {@link
+   * Enrollment#getAttributes()}. Since attribute values in the sms can be tracked entity type
+   * and/or program attributes they have to be split. Tracked entity type attributes go to the
+   * {@link TrackedEntity#getAttributes()} while program attributes go to the {@link
+   * Enrollment#getAttributes()}. The confusion in our tracker model is that a program attribute
+   * <em>can</em> also be a tracked entity attribute. This is why its legal to put such attributes
+   * into both attribute collections. This is what Capture app currently does. This only works in
+   * the Capture app as it can guarantee that tracker programs will always contain all the tracked
+   * entity type attributes of the programs tracked entity as its program attributes.
+   *
+   * @param smsAttributeValues all attribute values of an {@link EnrollmentSmsSubmission}
+   * @param programAttributes program attributes of the tracker program the {@link
+   *     EnrollmentSmsSubmission} is for
+   * @param existingAttributeValues program and tracked entity type attribute values of an existing
+   *     tracked entity
+   * @return enrollment (program) attributes for the tracker importer
+   */
+  @Nonnull
+  static List<Attribute> mapProgramAttributeValues(
+      @Nullable List<SmsAttributeValue> smsAttributeValues,
+      @Nonnull Set<String> programAttributes,
+      @Nonnull Set<String> existingAttributeValues) {
+    List<SmsAttributeValue> smsProgramAttributeValues =
+        emptyIfNull(smsAttributeValues).stream()
+            .filter(av -> programAttributes.contains(av.getAttribute().getUid()))
+            .toList();
+    Set<String> existingProgramAttributeValues =
+        existingAttributeValues.stream()
+            .filter(programAttributes::contains)
+            .collect(Collectors.toSet());
+    return mapAttributeValues(smsProgramAttributeValues, existingProgramAttributeValues);
+  }
+
+  @Nonnull
+  private static Event mapToEvent(
```

Comment:
I can see there are different `map` and some `mapTo`. Is it on purpose or we should just stick with one naming standard?

Reply:
The return type is not part of the method signature. That is why there are cases where I had to use a different name. I did try to use `map` for the package private methods that turn an sms submission into tracker objects. Other private helpers then got other names.